### PR TITLE
Show gem owners that are not using MFA when logged in as owner

### PR DESCRIPTION
**Pull request by lmansur**
As a response to recent gem exploits, we are seeing an increased desire to ensure all gem owners are using MFA.

This PR adds a way for gem owners to check if their fellow owners are using MFA. Regular users **can't** see this. 

The following gif shows the behavior when logged in as an owner and MFA is not enabled for everyone:

rubygem_owner_view

The following screenshot shows the behavior when logged in as an owner and MFA is enabled for everyone:

image

### CODE REVIEW

File: app/assets/javascripts/pages.js [hunk 1]
```diff
@@ -63,3 +63,11 @@ $('.stats__graph__gem__meter').each(function() {
   bar_width = $(this).data("bar_width");
   $(this).animate({ width: bar_width + '%' }, 700).removeClass('t-item--hidden');
 });
+
+//gem page
+$('.gem__users__mfa-text').on('click', function() {
```
Comment (eliotsykes) [inline app/assets/javascripts/pages.js]: Consider putting this into a `$(document).ready(...)` to ensure it runs after the DOM is loaded and the `.gem__users__mfa-text` elements are present.

File: test/unit/helpers/rubygems_helper_test.rb [hunk 2]
```diff
@@ -117,6 +117,21 @@ class RubygemsHelperTest < ActionView::TestCase
       assert_equal expected_links, links_to_owners(@rubygem)
       assert links_to_owners(@rubygem).html_safe?
     end
+
+    should "create links to gem owners without mfa" do
+      with_mfa = create(:user, mfa_level: "ui_and_api")
+      without_mfa = Array.new(2) { create(:user, mfa_level: "disabled") }
```
Comment (eliotsykes) [inline test/unit/helpers/rubygems_helper_test.rb]: Consider using FactoryBot's `create_list`:

```diff
-without_mfa = Array.new(2) { create(:user, mfa_level: "disabled") }
+without_mfa = create_list(:user, 2, mfa_level: "disabled")
```

File: app/views/rubygems/_gem_members.html.erb [hunk 3]
```diff
@@ -4,6 +4,17 @@
...
+    <% if rubygem.owned_by?(current_user) && rubygem.owners.without_mfa.present? %>
+      <span class="gem__users__mfa-text"><%= t '.not_using_mfa_warning_show' %></span>
+
+      <div class="gem__users__mfa-disabled t-item--hidden">
+        <span class="gem__users__mfa-text t-item--hidden"><%= t '.not_using_mfa_warning_hide' %></span>
+        <div class="gem__users">
+          <%= links_to_owners_without_mfa(rubygem) %>
+        </div>
+      </div>
+    <% end %>
```
Comment (eliotsykes) [inline app/views/rubygems/_gem_members.html.erb]: Consider showing a message along the lines of "All owners have MFA enabled" when that's the case.

**Comment by lmansur** [comment]
Thank you for your review, @eliotsykes .

Let me know what you think of the changes.

File: app/views/rubygems/_gem_members.html.erb (hunk 3 above)
Comment (lmansur) [reply app/views/rubygems/_gem_members.html.erb]: fixed in 53438a9

File: test/unit/helpers/rubygems_helper_test.rb (hunk 2 above)
Comment (lmansur) [reply test/unit/helpers/rubygems_helper_test.rb]: fixed in 6a6d3a3

File: app/assets/javascripts/pages.js (hunk 1 above)
Comment (lmansur) [reply app/assets/javascripts/pages.js]: fixed in d68917a

**Comment by geemus** [comment]
Looks good to me, thanks for taking this on!

**Comment by eliotsykes** [comment]
Thanks for building this @lmansur. The changes look good.

---

IIRC, currently any gem owner can silently and without confirmation add another user as an owner.  

If this is still true, what risks are there, if any, of the feature in this PR being abused?

One risk might be a malicious owner could add another user as an owner to determine their MFA settings. Is this a problem?

**Comment by lmansur** [comment]
I think it can be exploited.

However, I also think the benefit of having a way to ensure all owners are using MFA outweighs this risk.

I would love some more input into this matter.

**Comment by geemus** [comment]
It may be that being able to add people without confirmation is more a bug/coincidence than a feature, but that also seems way out of scope for this particular change.

Is there additional danger associated with knowing that someone doesn't have MFA? You would still need their password, and if you already had their password you would already be able to find out if they had MFA by attempting to use it. Given that, I'm not sure we are any worse off, but I know my security-mindedness is not the strongest. Please let me know if there is a case I'm not thinking about.

**Comment by lmansur** [comment]
I think it's more a case of specifically targeting owners that don't have MFA and not wasting time with those that do.

It's not ideal, but I wouldn't call that a blocker.

**Comment by lmansur** [comment]
Anything we can do to move this forward?

File: config/locales/es.yml [hunk 4]
```diff
@@ -300,8 +300,11 @@ es:
       header: "dependencias de %{title}"
     gem_members:
       authors_header: Autores
+      not_using_mfa_warning_show:
```
Comment (bronzdoc) [inline config/locales/es.yml]: `not_using_mfa_warning_show: "* Algunos propietarios no están usando MFA. Haga click para ver la lista completa."`

`not_using_mfa_warning_show: "* Los siguientes propietarios no están usando MFA.  Haga click para ocultar."`

`using_mfa_info: "* Todos los propietarios están usando MFA."`

File: config/locales/fr.yml [hunk 5]
```diff
@@ -290,8 +290,11 @@ fr:
       header: Dépendances de %{title}
     gem_members:
       authors_header: Auteurs
+      not_using_mfa_warning_show:
```
Comment (bronzdoc) [inline config/locales/fr.yml]: Maybe we should research and add notice text for other languages?

**Comment by bronzdoc** [comment]
@lmansur Thanks for the PR!

This seems like a nice feature It will help create awareness around not using MFA.

File: test/integration/yank_test.rb [hunk 6]
```diff
@@ -37,7 +37,7 @@ class YankTest < SystemTest
     assert page.has_content?("Yanked by")
 
     css = %(div.gem__users a[alt=#{@user.handle}])
-    assert page.has_css?(css, count: 2)
+    assert page.has_css?(css, count: 3)
```
[the remark this answers is not in the document]
Comment (lmansur) [reply test/integration/yank_test.rb]: This PR adds a row of owners not using MFA, similar to the original row of owners. They share CSS, so they ended up getting matched by the above test.

File: config/locales/fr.yml (hunk 5 above)
Comment (lmansur) [reply config/locales/fr.yml]: I'm open to contributions here. I'd rather not translate it directly using something like Google Translate

File: config/locales/es.yml (hunk 4 above)
Comment (lmansur) [reply config/locales/es.yml]: Thanks! Added in 5cab56c

**Comment by lmansur** [comment]
@bronzdoc I appreciate you taking the time to review this!

File: app/views/rubygems/_gem_members.html.erb [hunk 7]
```diff
@@ -4,6 +4,21 @@
     <div class="gem__users">
       <%= links_to_owners(rubygem) %>
     </div>
+
+    <% if  rubygem.owned_by?(current_user) %>
+      <% if rubygem.owners.without_mfa.present? %>
```
Comment (nateberkopec) [inline app/views/rubygems/_gem_members.html.erb]: .present? loads the entire relation (`SELECT * FROM USERS`). You don't actually want to do that here. Substitute `exists?` instead.

**Comment by mensfeld** [comment]
@geemus I believe it should also include the code for the data dumps export. This info is super valuabvle for probabilistic exploit analysis (cc @colby-swandale @dwradcliffe)

**Comment by indirect** [comment]
Thanks, this seems like a solid improvement!

**Comment by geemus** [comment]
Thanks!

**Comment by lmansur** [comment]
Thanks, everyone! :tada:

**Pull request merged**